llvm/include/llvm/Transforms/Utils/SYCLModuleSplit.h

```diff
@@ -0,0 +1,250 @@
+//===--------- SYCLModuleSplitter.h - split a module into callgraphs
+//-------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Functionality to split a module into call graphs. A callgraph here is a set
+// of entry points with all functions reachable from them via a call. The result
+// of the split is new modules containing corresponding callgraph.
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SYCL_MODULE_SPLIT_H
+#define LLVM_SYCL_MODULE_SPLIT_H
+
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Function.h"
+#include "llvm/Support/Error.h"
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+// TODO(maksimsab):
+//  * remove EntryPointGroup::Properties and maybe EntryPointGroup at all.
+//  * perhaps, ModuleDesc::IsTopLevel worth to be removed as well as #ifdef
+//  NDEBUG
+//  * Stub string values for ModuleDesc::Name don't look OK.
+//  * check GenXSPIRVWriterAdaptor comments
+//  * Maybe fix doxygen comments.
+
+namespace llvm {
+
+class Function;
+class Module;
+
+enum IRSplitMode {
+  SPLIT_PER_TU,     // one module per translation unit
+  SPLIT_PER_KERNEL, // one module per kernel
+  SPLIT_AUTO,       // automatically select split mode
```

Comment:
Note that both `auto` (which is `per-source` in 95-99% of cases) and `per-source` are also doing nested per-optional-kernel-feature split and we don't do such split in `none` mode.

I remember that it was confusing for someone even internally and I anticipate even more confusion in the upstream.
Perhaps, the commands should be more elaborate here, i.e. we should state that if _any_ kind of split is performed, then we also guarantee that every resulting module uses a unique set of optional features.

Reply:
I am preparing a supporting document and I will add details there. I will try to get that out before coming Monday. Thanks